content/md/posts/2023-07-08-testing.md

```diff
@@ -0,0 +1,150 @@
+
+To setup this test I need to initialize firebase and the routing for my application beforehand. 
+
+### Initialize firebase
+
+For information about how to setup firebase, look in the post called "Setting up the project part 2"
+
+### Initialize routing
+
+For information about how to setup routing, look in the post called "Setting up the project part 1"
+
+We can now write our test for registering a user. If the user is registered and logs in, then their account information should be updated to reflect them being the current user. 
+
+```clojure
+(t/deftest registered-user-login
+  (rf-test/run-test-async
+   (routes/init-routes!)
+   (core/firebase-init!)
+   (rf/dispatch-sync [::events/initialize-db])
+   (rf/dispatch [::events/login])
+   (rf-test/wait-for [::events/fetch-user] 
+    (let [user (rf/subscribe [::subs/user])
+          name (rf/subscribe [::subs/name])]
+      (t/are [x y] (= x y)
+        :registered @user
+        "han@skywalker.com" @name)))))
+```
+
+First, I initialize my routes, setup the database, and connect to firebase. After that, I login and wait for the user's information to be fetched. I then check whether the name of the user matches the credentials entered earlier and if the user is noted as registered.
+
+## Acceptance testing
+
+Acceptance testing is a bit different from the first two tests, as it tests things from the perspective of the user. You think of what the user would need to do to in order to do something, such as logging in, and you want to test whether what they see matches what should be there.
+
+For example, in my application when the user logs in they are taken to their dashboard and shown a welcome message. I want to see that when they reach the dashboard, they see the welcome message. To make this test, we need to set a few things up.
+
+Setup the directory for acceptance testing as directed in the post "Setting up the project part 2". After, create a file in test/acceptance called admin_login.clj. Here, we are going to put the test that we described above.
+
+Change the namespace so it appears like this. 
+```clojure
+(ns acceptance.admin-login-test
+  (:require  [clojure.test :as t]
+             [etaoin.api :as e]
+             [etaoin.keys :as k]
+             [acceptance.core :as c :refer [driver]])
+  (:import com.google.firebase.cloud.FirestoreClient))
+```
+
+Then, add the following code to the file.
+
+```clojure
+(def default-message
+  "Hello (Unknown). We're glad to see you.")
+
+(def admin-login-message
+  "Hello han@skywalker.com. Welcome to your dashboard. To make a new document, click the button to your right. To share and edit your existing documents, look at your documents below")
+
+(t/use-fixtures :once c/init-firebase) ;; Sets up firebase once for each test
+(t/use-fixtures :each c/with-screenshot) ;; Takes a screenshot at the end of every test
+
+(t/deftest message-test
+  (t/testing "When the admin user exists"
+    (doto driver
```

Comment:
Maybe we can replace this (just for the blog post) with a function with a nice meaningful name like `(login-and-open-dashboard-page)` or something like that...

Also perhaps remove the screenshot in the middle

Reply:
Ok. I'll take that in mind and edit all the tests so the names are more meaningful.